src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+        assert!((oob_data_len as usize) < (WRITE_BUFFER_SIZE-HEADER_SIZE), "too much oob data");
+        (WRITE_BUFFER_SIZE-HEADER_SIZE) - (oob_data_len as usize)
+    }
+
+    fn needs_fragmentation(data_len: usize, oob: &OsIpcOutOfBandMessage) -> bool {
+        let oob_size = if oob.needs_to_be_sent() { bincode::serde::serialized_size(oob) } else { 0 };
+        data_len > OsIpcSender::maximum_message_data_size(oob_size)
+    }
+
+    // An internal-use-only send method that sends just data, but doesn't
+    // fragment -- we use this for the secondary channel fragmentation
+    // creates.
+    fn send_unfragmented(&self, data: &[u8]) -> Result<(),WinError> {
+        assert!(data.len() < INVALID_HEADER_DATA_SIZE as usize);
+
+        dd2!("[c {:?}] writing unfragmented to (pid {}->{})", *self.handle, *CURRENT_PROCESS_ID,
+             try!(self.get_pipe_server_process_id()));
+
+        unsafe {
+            let header: [u32; 2] = [ data.len() as u32, 0 ];
+            let header_bytes: &[u8] = slice::from_raw_parts(header.as_ptr() as *const u8, HEADER_SIZE);
+            try!(write_buf(*self.handle, &header_bytes));
+            try!(write_buf(*self.handle, data));
+        }
+        dd2!("[c {:?}] sent unfragmented: {} bytes", *self.handle, data.len());
+        Ok(())
+    }
+
+    pub fn send(&self,
+                data: &[u8],
+                ports: Vec<OsIpcChannel>,
+                shared_memory_regions: Vec<OsIpcSharedMemory>)
+                -> Result<(),WinError>
+    {
+        // We limit the max size we can send here; we can fix this
+        // just by upping the header to be 2x u64 if we really want
+        // to.
+        assert!(data.len() < INVALID_HEADER_DATA_SIZE as usize);
+
+        let mut server_process_handle_opt: Option<WinHandle> = None;
+        let mut server_process_handle = || {
+            if server_process_handle_opt.is_none() {
+                server_process_handle_opt = Some(self.get_pipe_server_process_handle().unwrap());
+            }
+            **server_process_handle_opt.as_ref().unwrap()
+        };
+
+        let mut oob = OsIpcOutOfBandMessage::new();
+
+        for ref shmem in shared_memory_regions {
+            // shmem.handle, shmem.length
+            let remote_handle = try!(dup_handle_to_process(*shmem.handle, server_process_handle()));
+            oob.add_shmem(remote_handle, shmem.length);
+        }
+
+        for port in ports {
+            match port {
+                OsIpcChannel::Sender(mut s) => {
+                    let raw_remote_handle = try!(move_handle_to_process(s.handle.take(), server_process_handle()));
+                    oob.add_sender(raw_remote_handle);
+                },
+                OsIpcChannel::Receiver(mut r) => {
+                    if try!(r.prepare_for_transfer()) == false {
+                        panic!("Sending receiver with outstanding partial read buffer, noooooo!  What should even happen?");
```

Comment:
Yeah, I thought about this.. I sort of convinced myself that receivers/senders will only get sent right after they are created, before they're actually used.  However, for Windows at least, I do have a mitigation for the moving-receiver scenario.  I can send the process id of the intended target process, that the receiver can then compare.. if it differs, the receiver can move the handles over to itself *as long as the original target process still exists*.  That would help in some, but not all cases.

Cross-process locking is trivial, yeah. (Kernel handles also represent synchronization primitives, and they can be copied across processes.)